src/dune_rules/melange/melange_rules.ml

```diff
@@ -438,12 +438,27 @@ let setup_emit_js_rules ~dir_contents ~dir ~scope ~sctx mel =
   in
   let* compile_info = compile_info ~scope mel in
   let+ () =
-    let* requires_link =
-      Lib.Compile.requires_link compile_info
-      |> Memo.Lazy.force >>= Resolve.read_memo
+    let* requires_link_resolve =
+      Lib.Compile.requires_link compile_info |> Memo.Lazy.force
     in
-    setup_js_rules_libraries ~dir ~scope ~target_dir ~sctx ~requires_link ~mode
-      mel
+    match Resolve.is_ok requires_link_resolve with
+    | true ->
+      let* requires_link = Resolve.read_memo requires_link_resolve in
+      setup_js_rules_libraries ~dir ~scope ~target_dir ~sctx ~requires_link
+        ~mode mel
+    | false ->
+      let loc = mel.loc in
+      (* NOTE: for monorepo scenarios where single packages can be built
+       * separately (with [dune build -p <PKG>]), we can't eagerly resolve the
+       * link-time closure for the [melange.emit] stanza since melange.emit
+       * targets aren't public (i.e. part of a package.)
+       * In the cases where resolution fails, we still add dummy targets to
+       * give Dune a chance to faile if these targets end up attached to a
+       * package installation. *)
+      Super_context.add_rule sctx ~dir ~loc ~mode
+        (Action_builder.fail { fail = (fun () -> assert false) }
```

Comment:
You can raise the equivalent of `Resolve.read_memo` inside `fail`. You might need to add a function to `Resolve` to raise `Resolve.error` though.